Flex Compensation for Winch Kinematics

tobbelobb opened the pull request:
I meant to try and talk to you before creating a full PR, but here it is.

After a lot of effort I managed to post something on your Discourse:
https://klipper.discourse.group/t/flex-compensation-for-winch-kinematics/25031/3
Please answer here, your Discourse is unusable for me since I'm a new user.

It supports any purely translational cable robot configuration that has a central effector, including SpiderCam-like setups (all anchors high up), Slideprinter (2D robots), HP3-like setups (one high, several low anchors). Up to 26 anchors.

The "Print Logo" and "Print Squares" over at https://tobbelobb.github.io/hp-sim5/hp-sim/ now do flex compensation.
See https://gitlab.com/tobben/hangprinter-flex-compensation/ for the algorithm development.
See https://github.com/tobbelobb/klipper for the klipper implementation.
See https://github.com/tobbelobb/klipper/blob/master/docs/Config_Reference.md#cable-winch-flex-compensation for documentation on how to configure flex compensation.

<img width="1882" height="1241" alt="Screenshot from 2025-10-17 14-28-48" src="https://github.com/user-attachments/assets/1d456249-aeb7-4e23-8dfe-958d3438e829" />
<img width="1882" height="1241" alt="Screenshot from 2025-10-17 14-26-18" src="https://github.com/user-attachments/assets/2ead09be-e6fc-4265-89b1-88c4defd1e6f" />
<img width="1882" height="1241" alt="Screenshot from 2025-10-17 13-23-54" src="https://github.com/user-attachments/assets/6eb8a79d-e47f-42ad-a52b-bdc5ba1e7108" />
<img width="1882" height="1241" alt="Screenshot from 2025-10-17 13-19-29" src="https://github.com/user-attachments/assets/115391c3-f5ed-47d0-b0d3-a0d8a50285e2" />

tobbelobb committed:
feat: Flex compensation for Hangprinters

tobbelobb committed:
Removes some buggy cache stuff

tobbelobb committed:
Winch flex compensation: Prints optimal pretension at the origin

tobbelobb committed:
Initial integration of thikonov and qp flex compensation

tobbelobb committed:
Removes 'legacy' flex compensation algorithm

tobbelobb committed:
Updates Config_Reference.md with the behavior we're about to implement

tobbelobb committed:
Fix build errors

tobbelobb committed:
Replaces "hard" and "soft" with unified flex_compensation_algorithm

Also removes reference to "legacy" algorithm,
updates config/example-winch.cfg and adds more
explanation to config reference.

tobbelobb committed:
WIP: Integrate Thikonov and QP flex compensation in Winch kinematics

tobbelobb committed:
WIP: Compiles again

tobbelobb committed:
Corrects thikonov/tikhonov spelling err

tobbelobb committed:
Allow Tikhonov to output negative forces if requested

tobbelobb committed:
Fix typo

tobbelobb committed:
Make M666 F1 enable flex compensation without causing stepper timer issues

tobbelobb committed:
Just use the flex subtracted from absolute dist

tobbelobb committed:
Make M666 F1 produce real queue_step activity instead of just toggling a flag.

tobbelobb committed:
Builds a softer nullspace projector to tikhonov gradient descent doesn't crash

tobbelobb committed:
Makes qp the default flex compensation algorithm

tobbelobb committed:
Removes  from config file. Only enable/disable during runtime

tobbelobb committed:
Remove question from config file

tobbelobb committed:
Merge remote-tracking branch 'upstream/master'

tobbelobb committed:
Fixes stray old tikhonov default

tobbelobb committed:
Updates winch flex tests.

KevinOConnor commented:
Interesting, thanks.  I have some high-level comments (in no particular order).

There is a lot of dense math here that I wont be able to review nor maintain.  It's still possible that this code could reside in the main Klipper repository, but if that is the goal you would need to agree to maintain the code (answer questions, run tests, and make updates well into the future).  Alternatively, perhaps @dmbutyugin would be willing to review.

If you are looking to merge into the repo, the existing build test cases would need to pass.  Also, note that the directory `test/klippy/` is for black-box testing - it's okay to add unit tests but they should go in a new directory.

Similarly, if looking to merge into this repo, a signed-off-by line containing your real full name would be needed ( https://www.klipper3d.org/CONTRIBUTING.html#format-of-commit-messages ).  Any file copyrights added should also contain a full name.

Although I did not review the code, I get the impression that the new `kin_winch.c` code does a lot of processing on each "guess" of the iterative solver.  That sounds rather cpu intensive.  It's fine if that is what it takes, but there should be awareness of that if it is the case.

Note that there have been some other users of "cable winch" kinematics outside of the "hang printer" category.  At least one person is using that kinematics for extreme speed prints.  I'll followup on Discourse.

FYI, for this upstream repo, we basically never add old-style "M1234" type commands.  All new commands are expected to be in the new format (eg, `SET_CABLE_WINCH_KINEMATICS SOMEPARAM=123.4`).

Cheers,
-Kevin

tobbelobb commented:
Hi, thanks for the comments!

There's no rush to merge this now but your input is super useful, so I can work in that direction. We can look at it again when it's matured a little bit. I'm noting these todos:

 - Needs a person responsible. I'll do that once it's ready. If somebody else is willing to, I'd be happy to help them get up to speed with better docs and so on.
  - Line length, whitespace test cases, signed-off commit message. No problem.
  - Change F666 F0/1 to new format. No problem.
  - Some data regarding performance.
  - Testing. Move unit test to its own dir. Probably add black-box tests. Probably add plotting visual tests similar to what's in the [development repo](https://gitlab.com/tobben/hangprinter-flex-compensation/) (see "plotstuff.sh").
   - A mathematical formulation of what I'm checking for in plots would be nice, so we can automate the "plotstuff" tests.
     * No force spikes.
     * Smooth force changes in all directions from all positions with all supported setups.
     * Not too large fraction of the build volume getting forces clamped in normal setups. For some definition of "too large" and "normal".
     * Gravity being supported.
     * The two algorithms giving fairly similar results at all points except near edges where they're expected to differ
  - Document effect on real prints. I don't have the setup to do this right now. Loosing my workshop is what has given me so much time to work on software stuff, so I'll probably need help on this one, and it might take a lot of time to test all supported setups. I guess that's ok, as long as interested potential users can find my fork in the meantime.

### Some Notes on Performance
I don't expect performance will be a problem, but we'll see. The default QP algorithm and the other one ("Tikhonov", which uses gradient descent) have different performance properties. I'll try and get some real numbers today.

Does Klipper have a target machine or a minimum requirement on the users CPU?

#### Appendix Notes on Performance
You're right to be a bit suspicious about the numerical solving, it needs to be tightly measured/tuned/controlled, especially since performance/solve times varies in different areas/volumes:
    * Easily reachable, vs
    * Near singularities vs,
    * Past singularities (physically unreachable positions).

To my knowledge the numerical iterations is what it takes to solve this (most often underdetermined) non-linear optimization problem in a stable, reliable, and general manner. Some setups have (potentially a lot of) closed form solutions but I don't think they're worth implementing individually.

It's possible to get forces faster by trading away perfection for speed. It depends on how close to "optimial" we want/need the (already correct) forces to be. I've used different definitions of "optimal" for the two algorithms:
#### QP forces are:
  - As close to min as possible, from above
  - Never below min
  - Never negative
  - Never above max
  - Always physically perfect (forces balance out perfectly, even near singularities)
  
  #### Tikhonov forces are:
   - As close to min as possble
   - Preferrably not below min, and if below, still very close to min
   - Never negative if the configured min force is positive
   - Allowed to go negative if the user configured a negative min force
   - Never above max
   - Close to physically perfect, but not if that means sudden force changes or very high total force

tobbelobb commented:
Ok I did some preliminary performace checks on my desktop machine. I did it like [this](https://github.com/tobbelobb/klipper/commit/27b63cbf6c452152bf257aee7474896f4b3d788c).

Results in
[klipper_winch_flex_perf_data.log](https://github.com/user-attachments/files/23119818/klipper_winch_flex_perf_data.log)

I used this gcode as the example object:
[Hangprinter_logo6.gcode.log](https://github.com/user-attachments/files/23119871/Hangprinter_logo6.gcode.log)
It's a ca 800 mm X-wide and ca 940 mm Y-high and 1.8 mm Z-thick (three layers). 2050 s print time.

## Main Result in the Slideprinter (three low anchors) Case
Out of 2050 s print time, the `compute_flex` function consumed ~5.0 s when QP was used, and ~8.5 s when Tikhonov was used. They were called 25 million times, so each call was around .2-.3 ns.

```
QP ignore_gravity: false, ignore_pretension: false:
ERROR:root:b'compute_flex calls=24883580, total=4999.317 ms, avg=0.000201 ms'
Tikhonov ignore_gravity: false, ignore_pretension: false:
ERROR:root:b'compute_flex calls=25092895, total=8214.606 ms, avg=0.000327 ms'
```

## Comments About Number of Lines Effect on Performance
We currently calculate forces in all directions each time we calculate one single line length. We could optimize this. So for a machine with 3 lines we could cut the number of calls by 1/3 with some caching mechanism.

Results also depend a lot on the number of anchors and the max/min force limits we configure. I've used `winch_min_force: 3N`, `winch_max_force: 170N` on all anchors.

Increasing the number of anchors from 3 to 4 (standard HP3 setup) gave:
```
QP ignore_gravity: false, ignore_pretension: false:
ERROR:root:b'compute_flex calls=26639211, total=6580.517 ms, avg=0.000247 ms'
Tikhonov ignore_gravity: false, ignore_pretension: false:
ERROR:root:b'compute_flex calls=26848825, total=38785.034 ms, avg=0.001445 ms'
```

We see that Tikhonov does not scale as well as the default QP. This got me curious so I added another (high up) anchor and got:
```
QP:
ERROR:root:b'compute_flex calls=30377819, total=9812.713 ms, avg=0.000323 ms'
Tikhonov:
ERROR:root:b'compute_flex calls=31203995, total=65175.208 ms, avg=0.002089 ms'
```

The results are roughly the same if the 5 anchors are arranged like HP5 will have them, with 4 low and 1 high anchor:
```
QP:
ERROR:root:b'compute_flex calls=35115829, total=10850.151 ms, avg=0.000309 ms'
Tikhonov:
ERROR:root:b'compute_flex calls=37424930, total=76917.615 ms, avg=0.002055 ms'
```

We can conclude that the current Tikhonov algorithm is too slow for machines with more than 3 anchors. They should probably stick to the default QP algorithm.

## Comments About Whether this Will Work On Raspberry Pi Zero2W
My desktop cpu: 12th Gen Intel(R) Core(TM) i7-12700K

The Pi Zero2W has a Quad-core 1GHz Cortex-A53 which is a lot slower. But if we can fit the matrices in the Zero's cache (which I think we can, the matrices are small), then 10s on my desktop will probably expand to ca 300 s on the Pi Zero2W (assuming we run single threaded). That's just my guess.

Is that ok, or would 300 s of flex calculation on a 2050 s print be too slow?

tobbelobb committed:
Prepares update of winch forward transform by adding some reference code and docs

tobbelobb committed:
Implements quadratic algo forward transform for winch

tobbelobb committed:
Adds reference test harness for quadratic forward transform algo for winch

tobbelobb committed:
Fixes bug in winch forward transform

tobbelobb committed:
Updated reference motorstepstocartesiantest_quadratic.cpp

tobbelobb committed:
WIP: Port quadratic forward kinematics to klipper

run with
python ai_docs/klipper_forward_harness.py

tobbelobb committed:
Updates winch forward transform test datasets

tobbelobb committed:
Simplifies winch kinematics

Makes behavior more similar to RRF:
 - No immediate movement upon enabling flex compensation
 - Count flex compensation in absolute terms, not relative to the flex at the origin
 - Include a configurable mechanical_advantage
 - Removes old messy attempt at forward transform (calc_position).

tobbelobb committed:
Brings back the immediate flex transition

tobbelobb committed:
Updates winch forward transform reference

tobbelobb committed:
Introduces buildup_factor and uses distances that are relative to the origin.

The buildup compensation is not applied yet, it's just prepared for.

tobbelobb committed:
Adds winch buildup compensation feature

tobbelobb committed:
Implements Hybrid Halley (winch forward transform)

Main logic in kin_winch.c

tobbelobb committed:
Removes unused code

tobbelobb committed:
Test calc_position

tobbelobb committed:
Winch forward transform test is green

See ai_docs/test.sh

tobbelobb committed:
Documents the winch_buildup_factor and includes it in test

tobbelobb committed:
Clarify that mechanical_advantage should be manually folded into rotation_distance

tobbelobb committed:
Don't set _last_forward to None when we know it's not None.

tobbelobb committed:
Fixes absolute/relative error

tobbelobb committed:
Uses relative distances in flex enable target_length

tobbelobb committed:
Activate flex compensation when M666 F1 is sent. Not before.

tobbelobb committed:
Adds placeholder support for M569.3

tobbelobb committed:
Adds placeholder support for M569.4

tobbelobb committed:
Make M569.3 and M569.4 P addresses explicit in .cfg

tobbelobb committed:
Removes M569 P default of 40. Makes stepper name default instead
